### ISSUE TITLE
Inconsistent editor focus behaviour when opening a document from project tree

### ISSUE THREAD

**Issue by tmarplatt** [vkbo/novelWriter#1368]
**The problem**

From the project tree, selecting a file and opening it with either:

- Enter key
- via menu bar: Document > Open Document 
- Ctrl+O shortcut

None does focus the editor afterwards, as opposed to other methods (i.e. double click; context menu "Open Document").

**The expectation**

All methods used above should focus the editor once the doument is loaded.

**Comment by tmarplatt** [comment]
I don't know if it's the same issue, but a similar thing happens in Novel tree and Outline views.

**Comment by vkbo** [comment]
This is not a bug. It is intentional.

The problem here is to satisfy two contradictory expectations. Some people want to be able to flip through the documents by using the arrow keys and hitting one of the shortcut (`Enter` or `Ctrl+O`), but don't want to have to use the keyboard to re-focus the project tree again for each one. This means they have to manually switch focus when they do open the document they want to edit. Those two shortcuts can only work one way or the other, not both.

Since the `Ctrl+O` just activates the menu entry action, the menu entry behaves the same way.

When double-clicking, focus doesn't matter, so it does change focus, which is the only case where you actually can get both outcomes since switching focus doesn't hinder you double-clicking another item.

**Comment by tmarplatt** [comment]
But if the expectations are contradictory it's only because they originate in the same set of controls. In general, double-clicking and hitting Enter represent synonymous input, it comes with primary user expectation.

I wouldn't mind if, say, the tree's context menu had an **Open Document** entry above an **Open and Edit Document** entry. Now you satisfy both expectations by allowing two distinct set of controls. What follows is figuring out their alias shortcuts at the same time as which one is default (bound to Enter/double-click).

By the way judging from your reply I take you aren't aware of the actual Novel tree and Outline bugs (both seem to be the same) so I'm going to submit that in another ticket.

**Comment by vkbo** [comment]
> But if the expectations are contradictory it's only because they originate in the same set of controls. In general, double-clicking and hitting Enter represent synonymous input, it comes with primary user expectation.

The original behaviour was for the focus switched to the editor when opening a document for editing. That received user complaints due to the need to switch focus back if they were just "browsing". So the keyboard shortcut implementation was changed. The mouse interaction is a completely different code path, and was left alone.

I can switch the double-click to behave the same way, but for a mouse-user I doubt the difference matters. Switching everything to the double-click behaviour will just invite back the same complaints as before, so I'm not going to do that at this time.

> I wouldn't mind if, say, the tree's context menu had an **Open Document** entry above an **Open and Edit Document** entry. Now you satisfy both expectations by allowing two distinct set of controls. What follows is figuring out their alias shortcuts at the same time as which one is default (bound to Enter/double-click).

That sounds overly complicated and confusing.

> By the way judging from your reply I take you aren't aware of the actual Novel tree and Outline bugs (both seem to be the same) so I'm going to submit that in another ticket.

I am, but they are unrelated to this. They need a new issue ticket, so thanks for doing that.

**Comment by vkbo** [comment]
> But if the expectations are contradictory it's only because they originate in the same set of controls. In general, double-clicking and hitting Enter represent synonymous input, it comes with primary user expectation.

The original behaviour was for the focus to switch to the editor when opening a document for editing. That received user complaints due to the need to switch focus back if they were just "browsing". So the keyboard shortcut implementation was changed. The mouse interaction is a completely different code path, and was left alone.

I can switch the double-click to behave the same way, but for a mouse-user I doubt the difference matters. Switching everything to the double-click behaviour will just invite back the same complaints as before, so I'm not going to do that at this time.

> I wouldn't mind if, say, the tree's context menu had an **Open Document** entry above an **Open and Edit Document** entry. Now you satisfy both expectations by allowing two distinct set of controls. What follows is figuring out their alias shortcuts at the same time as which one is default (bound to Enter/double-click).

That sounds overly complicated and confusing.

> By the way judging from your reply I take you aren't aware of the actual Novel tree and Outline bugs (both seem to be the same) so I'm going to submit that in another ticket.

I am, but they are unrelated to this. They need a new issue ticket, so thanks for doing that.

**Comment by vkbo** [comment]
An option is to add a modifier to the `Enter` shortcut, say `Ctrl+Enter`, to both open and switch focus. That would at least make it easier for keyboard users, and does not have to interfere with the work flow of mouse users.

I would not add a menu entry specifically for this, so it would be a feature that would require that you check the documentation. There are a handful of suggested cases where it is desirable to add a shortcut, but not natural to add a GUI element. I have not decided how to best document such features. The documentation has an overview of all shortcuts, but it may be useful to have a cheat sheet somewhere as well. I'm a little reluctant to add a GUI overview of shortcuts, as we have restrictions on how much text we can translate on Crowdin. But it may be the best option regardless.

**Comment by tmarplatt** [comment]
> There are a handful of suggested cases where it is desirable to add a shortcut, but not natural to add a GUI element. I have not decided how to best document such features.

The oldschool way of doing that is via a Hints and Tips dialog at program launch. 😄

> I'm a little reluctant to add a GUI overview of shortcuts, as we have restrictions on how much text we can translate on Crowdin.

You can safely reuse strings here, surely? The same strings from menus and/or dialog titles.

**Comment by vkbo** [comment]
> The oldschool way of doing that is via a Hints and Tips dialog at program launch. :smile:

Yeah, I plan to rewrite the shortcuts page in the docs, and possibly link it from the Help menu directly.

> You can safely reuse strings here, surely? The same strings from menus and/or dialog titles.

The strings are per-component, so it depends on how I implement it. If I were to show the keybindings in a GUI, I would define them all in a single data class somewhere, and extract that info both in the overview and in the menus. It would allow for a decent amount of reuse. Maybe enough that it would be fine.

**Comment by tmarplatt** [comment]
Related, I noticed that adding an item does not imply focusing the project tree. That one catches me by surprise every time. I am not sure why you would want to keep focus in the editor/viewer if you're adding an item, which most of the time you'd want to edit/rearrange/open right away.

The project tree even selects the new item. Yet `Ctrl+O` does not open it.

**Comment by vkbo** [comment]
The project tree was redesigned quite a bit in 2.0, so there are still issues with the rewrite. Like the one you already identified. If you find more, just report them and I'll look into it.

I usually make my changes to the project tree using the mouse, so I notice those. But I don't often use the keyboard shortcuts for those specific tasks, so issues there I don't notice unless I explicitly test for them.

**Comment by vkbo** [comment]
I don't think I will change this behaviour. When I double-click an item in the tree, I would like to be able to just start typing where I left off, and where the cursor is already repositioned to. However, when I use the arrow keys in the tree and press Enter to see the content, I don't want to have to manually switch focus back.

The other issue from this thread has been fixed.
